File: algorithm/kotlin/src/main/kotlin/P68645.kt
class P68645 {
    fun solution(n: Int): IntArray {
        val snail = Array(n) { i -> IntArray(i + 1) { 0 } }


        var x = 0
        var y = 0
        var angle = 0

        var count = 1
        val dx = intArrayOf(0, 1, -1)
        val dy = intArrayOf(1, 0, -1)
        
        val until = n * (n + 1) / 2
        
        for (i in 1 .. until) {
            
            snail[y][x] = count
            
            val nx = x + dx[angle]
            val ny = y + dy[angle]
            count++

            if (0 <= nx && nx < n && 0 <= ny && ny < n && snail[ny][nx] == 0) {
                x = nx
                y = ny
            } else {
                angle = (angle + 1) % 3
                y += dy[angle]
                x += dx[angle]
            }

        }

        val list = mutableListOf<Int>()
        for (i in 0 until n) {
            for (j in 0..i) {
                list.add(snail[i][j])
            }
        }
        return list.toIntArray()
    }
}

fun main() {
    val p68645 = P68645()
    println(p68645.solution(3))
}

File: algorithm/python/P68645.py
def solution(n):

    # n * n 2차원 배열 생성
    snail = [[0] * i for i in range(1, n + 1)]

    # 반 시계 방향의 나선형 형태로 배열을 채워나간다.
    dx = [0, 1, -1]
    dy = [1, 0, -1]

    x = y = angle = 0
    cnt = 1

    size = (n + 1) * n // 2 # len(snail)
    while cnt <= size:
        snail[y][x] = cnt
        ny = y + dy[angle]
        nx = x + dx[angle]
        cnt += 1

        if 0 <= ny < n and 0 <= nx <= ny and snail[ny][nx] == 0:
            y,x = ny,nx
        else:
            angle = (angle + 1) % 3
            y += dy[angle]
            x += dx[angle]

    return [i for j in snail for i in j]